File: Python/LoopStartNode.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def find_loop_start(head: ListNode) -> ListNode:
    if not head:
        return None

    slow = head
    fast = head

    # Detect loop
    loop_exists = False
    while fast and fast.next:
        slow = slow.next
        fast = fast.next.next
        if slow == fast:
            loop_exists = True
            break

    if not loop_exists:
        return None

    # Find starting node of loop
    slow = head
    while slow != fast:
        slow = slow.next
        fast = fast.next

    return slow

# Example
head = ListNode(1, ListNode(2, ListNode(3, ListNode(4, ListNode(5)))))
head.next.next.next.next.next = head.next  # create loop at node with value 2

loop_start = find_loop_start(head)
print(loop_start.val if loop_start else "No loop")  # Output: 2

File: JAVA/LoopStartNode.java
class ListNode {
    int val;
    ListNode next;
    ListNode(int val) { this.val = val; this.next = null; }
}

public class LoopStartNode {

    public static ListNode findLoopStart(ListNode head) {
        if (head == null) return null;

        ListNode slow = head;
        ListNode fast = head;

        // Detect loop
        boolean loopExists = false;
        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
            if (slow == fast) {
                loopExists = true;
                break;
            }
        }

        if (!loopExists) return null;

        // Find starting node of loop
        slow = head;
        while (slow != fast) {
            slow = slow.next;
            fast = fast.next;
        }

        return slow; // start of loop
    }

    public static void main(String[] args) {
        ListNode head = new ListNode(1);
        head.next = new ListNode(2);
        head.next.next = new ListNode(3);
        head.next.next.next = new ListNode(4);
        head.next.next.next.next = new ListNode(5);
        head.next.next.next.next.next = head.next; // create loop at node with value 2

        ListNode loopStart = findLoopStart(head);
        System.out.println(loopStart != null ? loopStart.val : "No loop"); // Output: 2
    }
}
